### DISCUSSION TITLE
Implement Meilisearch for search engine

### DISCUSSION THREAD

**Discussion by ppamorim** [ThatGuySam/doesitarm#894] [Ideas]
Hi!

I been analysing the source code of the project and I noticed that the search functionality used is done all in the browser. Would you like to migrate it to a full text search like `Meilisearch`? I could help you to implement that in the project.

I can see this code here: This can be very expensive to process, in some situations it can take up to 1 second on my M1 Pro: https://github.com/ThatGuySam/doesitarm/blob/master/components/search.vue#L556

We can either use the self hosted version running in the same server or used the cloud solution. It's up to you.

Thank you

**Comment by ThatGuySam** [comment]
Thanks for the offer! 

One of the engineering goals I have for Does It ARM is to stay as static as possible, but it's getting to be tricky to load 3k apps into a browser. 

If we can keep the app as static as possible it won't be susceptible to computing providers being sold and/or shut down which doesn't always happen but I like to account for. 

My questions would be: 
- Can Meilisearch be implemented without a Database? 
- What is the infrastructure required for Self-hosted? 
- Can it be implemented while maintaining the current UI?

**Comment by ppamorim** [comment]
Hi @ThatGuySam, thank you for the answer! Answering some of the questions:

 **- Can Meilisearch be implemented without a Database?** 
 
 Meilisearch creates the database for you, it uses LMDB (Lightning Memory-Mapped Database) internally. You only store that in a disk on AWS if you want. Since the payload if small I believe that restarting the service and reindexing could be as fast as loading the snapshot.
 
  **- What is the infrastructure required for Self-hosted?** 
  
  You can read that here: https://docs.meilisearch.com/faq.html#what-are-the-recommended-requirements-for-hosting-a-meilisearch-instance
For this database, I would suspect that a single core machine + 1GB could be testable and very I believe that it can handle without problems.
Meilisearch also offers the cloud service, but it's on beta at the moment.

**Can it be implemented while maintaining the current UI?**

Yes we can, there is no need to use any specific UI for that, we can use the JS library for that: https://github.com/meilisearch/meilisearch-js

There is a series of tutorials explaining how to deploy the server: 

 AWS: https://docs.meilisearch.com/learn/cookbooks/aws.html
 GCP: https://docs.meilisearch.com/learn/cookbooks/gcp.html
 Kubernetes: https://github.com/meilisearch/meilisearch-kubernetes
 
 If you want to test it locally, you can try this: https://docs.meilisearch.com/learn/getting_started/quick_start.html#step-1-setup-and-installation

**Comment by ThatGuySam** [comment]
This looks like it will be decent amount of setup so let's convert this into a discussion

**Comment by ThatGuySam** [comment]
It looks like Laravel uses Meilisearch as well: 
https://laravel-news.com/laravel-scout-meilisearch

It looks there might be a way to run it on a Netlify Serverless function which would be ideal since we're already on Netlify, but we may need to setup a separate process to create the index
https://www.christopherbiscardi.com/running-meilisearch-in-a-lambda

**Comment by ThatGuySam** [comment]
Looks like there is a library called Stork that implements search with Web Assembly so it's optimized Static and JAMStack sites: 
https://github.com/jameslittle230/stork

This would be much easier to implement for Does It ARM and we can always switch over to Meilisearch if they introduce an official fully static implementation.

**Comment by ThatGuySam** [comment]
@ppamorim New Search is now implemented! 

You can try it out here: 
https://doesitarm.com/

Search is fully static so the browser doesn't have to wait for an indexing server and results are near-instant since there's no round trip for the requests.
